server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +270,181 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+            return err;
+        }
+    }
+    // in case we still don't have a proper device, use the default device
+    if (*device == paNoDevice) {
+        *device = defaultDevice;
+        if (*device != paNoDevice)
+            fprintf(stdout, "Selecting default system %s device\n", deviceType);
+    }
+    return paNoError;
+}
+
+void SC_PortAudioDriver::SelectMatchingPaDevice(int* matchingDevice, int* knownDevice,
+                                                IOType matchingDeviceType) const {
+    if (*matchingDevice != paNoDevice || *knownDevice == paNoDevice)
+        return;
+
+    const PaHostApiInfo* apiInfo;
+    apiInfo = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*knownDevice)->hostApi);
+
+    bool isInput;
+    if (matchingDeviceType == IOType::Input)
+        isInput = true;
+    else if (matchingDeviceType == IOType::Output)
+        isInput = false;
+
+    if (isInput) {
+        *matchingDevice = apiInfo->defaultInputDevice;
+    } else {
+        *matchingDevice = apiInfo->defaultOutputDevice;
+    }
+    if (*matchingDevice != paNoDevice)
+        fprintf(stdout, "Selecting default %s %s device\n", apiInfo->name, (isInput ? "input" : "output"));
+}
+
+// this function will select default PA devices if they are not defined
+// it will also try to check for some configuration problems
+// numIns, numOuts and sampleRate are only the requested values, can change later
+PaError SC_PortAudioDriver::CheckPaDevices(int* inDevice, int* outDevice, int numIns, int numOuts,
+                                           double sampleRate) const {
+    if (numIns && !numOuts) {
+        *outDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(
+            inDevice, sampleRate, (*inDevice != paNoDevice) ? Pa_GetDeviceInfo(*inDevice)->maxInputChannels : 0,
+            Pa_GetDefaultInputDevice(),
+            [](PaStreamParameters& params, double sr) { return Pa_IsFormatSupported(&params, nullptr, sr); }, "input");
+    } else if (!numIns && numOuts) {
+        *inDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(
+            outDevice, sampleRate, (*outDevice != paNoDevice) ? Pa_GetDeviceInfo(*outDevice)->maxOutputChannels : 0,
+            Pa_GetDefaultOutputDevice(),
+            [](PaStreamParameters& params, double sr) { return Pa_IsFormatSupported(nullptr, &params, sr); }, "output");
+    } else if (numIns && numOuts) {
+        // if one device is specified, let's try to open another one on matching api
+        SelectMatchingPaDevice(inDevice, outDevice, IOType::Input);
+        SelectMatchingPaDevice(outDevice, inDevice, IOType::Output);
+
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice
+            && Pa_GetDeviceInfo(*inDevice)->hostApi != Pa_GetDeviceInfo(*outDevice)->hostApi) {
```

Comment:
(suggestion) consider assigning this entire condition to a named bool before using it in the `if`, it could read clearer:

```cpp
bool apisAreDifferent = *inDevice != paNoDevice && *outDevice != paNoDevice
            && Pa_GetDeviceInfo(*inDevice)->hostApi != Pa_GetDeviceInfo(*outDevice)->hostApi;
if (apisAreDifferent) {
    ...
```